playground/didchat2/src/lib.rs

```diff
@@ -0,0 +1,240 @@
+        Ok(format!("Existing user already signed in."))
+    }
+}
+
+fn doc() -> Result<String, std::io::Error> {
+    use did_key::DIDCore;
+    use did_key::KeyMaterial;
+    
+    // 1. Read seed from file
+    let seed = std::fs::read(".didchat/seed").unwrap();
+
+    // 2. Transform seed to a did-document
+    let public_private_keypair = did_key::Ed25519KeyPair::from_seed(&seed);
+    let public_only_keypair = did_key::from_existing_key::<did_key::Ed25519KeyPair>(&public_private_keypair.public_key_bytes(), None);
+    let doc = public_only_keypair.get_did_document(did_key::CONFIG_LD_PUBLIC);
+
+    // 3. Serialize did-document to json
+    let json = serde_json::to_string_pretty(&doc).unwrap();
+    Ok(format!("{}", json))
+}
+
+fn did() -> Result<String, std::io::Error> {
+    use did_key::DIDCore;
+
+    // 1. Read seed from file
+    let seed = std::fs::read(".didchat/seed").unwrap();
+
+    // 2. Transform seed to a did
+    let keypair = did_key::Ed25519KeyPair::from_seed(&seed);
+
+    let diddoc: did_key::Document = keypair.get_did_document(did_key::CONFIG_LD_PUBLIC);
+    let did = diddoc.id;
+
+    // 3. Print did
+    Ok(format!("{}", did))
+}
+
+fn add(alias: &str, did: &str) -> Result<String, std::io::Error> {
+    use std::io::Write;
+
+    if !std::fs::metadata(".didchat/dids/").is_ok() {
+        std::fs::create_dir_all(".didchat/dids/").unwrap_or_default();
+    }
+
+    let friend_path = format!(".didchat/dids/{}.did", alias);
+    let mut file = std::fs::File::create(friend_path).unwrap();
+    file.write(did.as_bytes()).unwrap();
+
+    Ok(format!(".didchat/dids/{}.did", alias))
+}
+
+fn write(to: &str, message: &str) -> Result<String, std::io::Error> {
+    use did_key::Ecdh;
+    use did_key::KeyMaterial;
+    use did_key::DIDCore;
+
+    // 1. Read from and to x25519 keys from file
+    let from_seed = std::fs::read(".didchat/seed").unwrap(); // @unwrap!!
+    let from_key = did_key::Ed25519KeyPair::from_seed(&from_seed);
+    let from_did = from_key.get_did_document(did_key::CONFIG_LD_PUBLIC).id;
+    let from_key = from_key.get_x25519();
+    
+    let to_did = std::fs::read_to_string(format!(".didchat/dids/{}.did", to)).unwrap(); // @unwrap!!
+    let to_key = did_key::resolve(&to_did).unwrap(); // @unwrap!!
+    let to_key = did_key::Ed25519KeyPair::from_public_key(&to_key.public_key_bytes());
+    let to_key = to_key.get_x25519();
+
+    // 2. Generate elliptic-curve-diffie-hellmann (ecdh) shared secret
+    let shared_secret = from_key.key_exchange(&to_key);
+
+    // 3. Construct didcomm message
+    let to_vec = vec![&to_did[..]];
+    
+    let message = didcomm_rs::Message::new()
+        .from(&from_did)
+        .to(&to_vec[..])
+        .timed(Some(3600))
+        .body(message.as_bytes())
+        .as_jwe(&didcomm_rs::crypto::CryptoAlgorithm::XC20P);
+
+    // 4. Seal/encrypt message using shared secret
+    let sealed_message = message
+        .seal(&shared_secret)
+        .unwrap(); // @unwrap!!
```

Comment:
Unwrap after seal? How does that work?

Reply:
Haha. They are unrelated. This is the signature of the `seal()`-function:
```rs
 pub fn seal(self, ek: &[u8]) -> Result<String, Error>
```
The result of the `seal()`-function is wrapped inside a `Result`-container, so I have to `unwrap()` to get the actualt result. As @marni has mentioned, using `unwrap()` everywhere is an anti-pattern, which reflects my level of Rust-competence. I think the amount of `.unwrap()` in the code, is inversely proportional to my Rust-skills :laughing:

Reply:
`.unwrap()` is an error-handling pattern. There are many ways to handle errors in Rust. `.unwrap()` is the lowest cost option for beginners, but also the most dangerous and useless one, as it does not provide much extra information to the user of the software, if something goes wrong.

One good thing about `.unwrap()`, is that it communicates well to other developers, that you spent 0 calories considering what should happen, if an error occurs.